File: internal/domain/github_app.go
package domain

import (
	"fmt"
	"strings"
	"time"
)

type OrgGitHubApp struct {
	OrgID            string
	AppID            string
	Slug             string
	ClientID         string
	ClientSecretRef  string
	PEMRef           string
	WebhookSecretRef string
	// OwnerType is "user" when the App was registered under a personal
	// GitHub account, "org" when registered under an organization. Captured
	// from the signed registration state at callback time (not a fresh query
	// param) and surfaced through the status endpoint so Setup/Settings can
	// seed the "App account type" summary instead of re-defaulting to
	// "Personal account" on every page load. Mirrors the
	// org_github_apps.owner_type column (NOT NULL DEFAULT 'user').
	OwnerType          string
	RegisteredAt       time.Time
	RegisteredByUserID string
	Active             bool
	// BotUserID is the numeric GitHub *user-account* id of the App's bot
	// (the account "<slug>[bot]"), fetched best-effort at registration via
	// GET /users/<slug>[bot]. It is NOT the App id — the App id and the bot
	// user id are distinct numbers. 0 means unknown (an App registered before
	// this column existed, or a registration-time fetch failure); the commit
	// identity then falls back to the plain "<slug>[bot]@users.noreply.github.com"
	// form. When known, the resolver builds the numeric-id noreply email
	// "<bot_user_id>+<slug>[bot]@users.noreply.github.com", the only form that
	// links a bot's commits to its account on GitHub's contribution graph
	// (TFAC-474). Mirrors org_github_apps.bot_user_id (nullable → 0).
	BotUserID int64
}

// NormalizedOwnerType returns the App's owner_type, folding an unset value
// to "user" so the store never writes an empty string into the NOT NULL
// owner_type column (whose DB default is likewise "user"). The registration
// callback always supplies a validated "user"/"org"; this guards any other
// caller that constructs the struct without setting the field.
func (a OrgGitHubApp) NormalizedOwnerType() string {
	if a.OwnerType == "" {
		return "user"
	}
	return a.OwnerType
}

// OrgGitHubAppInstallation is one place the org's App is installed —
// a GitHub account (org or user) that ran the install flow. Mirrored
// from GitHub via the installation webhook (multi mode) or API
// backfill (local mode). One App can have many installations.
type OrgGitHubAppInstallation struct {
	InstallationID string
	OrgID          string
	AccountType    string
	// AccountID is GitHub's numeric id for the account the App is installed
	// on, in its text form (the convention entities.source_id and
	// InstallationID already follow). Unlike AccountLogin it survives a
	// rename, so it — not the login — is what credential resolution matches
	// on when both sides know it. "" means unknown: a row mirrored before
	// this was captured, which the next reconcile or webhook fills in.
	AccountID    string
	AccountLogin string
	// GitHubHost is the GitHub deployment this installation lives on — the
	// org's github_base_url normalized the way every other GitHub host key is
	// (db.EffectiveGitHubHost), so the same GitHub is the same string here as
	// in user_github_identities. It is on the row because an installation id is
	// unique per deployment and not universally: a self-host aggregating orgs
	// across two GHES instances can hold the same numeric id twice, meaning two
	// unrelated installations, and telling them apart through a join to the
	// owning org's settings is a join too many for any comparison that spans
	// orgs. "" on a struct handed to a writer means "resolve it" — the store
	// folds it to the public host, which is what an org with no configured base
	// URL is on — so reads never see one.
	GitHubHost  string
	InstalledAt time.Time
	// SuspendedAt is when the account owner suspended this installation, zero
	// when it is not suspended. A suspension is not an uninstall: the grant
	// survives, the row stays active (removed_at is untouched), and the
	// installation resumes on unsuspend — but every token minted from it is
	// refused while it lasts, so a suspended installation must be
	// distinguishable from a working one wherever the mirror is read.
	SuspendedAt time.Time
	// SuspendedBy is the GitHub login that suspended the installation, from the
	// payload's suspended_by user object. Display provenance only — nothing
	// resolves on it — and "" when the source named no one.
	SuspendedBy string
	// RepositorySelection is RepositorySelectionAll or
	// RepositorySelectionSelected — whether the grant is every repository on the
	// account or an enumerated set. It is what says whether scope drift is even
	// POSSIBLE for this installation: an 'all' grant reaches everything the
	// account owns, so a tracked repository on that account cannot sit outside
	// it, and "no drift found" there means something different from "no drift
	// found" under a selective grant.
	//
	// "" means unknown — a row mirrored before this was captured — and takes the
	// account id's fill-in-only write rule rather than the login's overwrite
	// one, because a writer that omits it is a writer that did not look.
	RepositorySelection string
}

// The two values GitHub reports for an installation's repository_selection.
const (
	RepositorySelectionAll      = "all"
	RepositorySelectionSelected = "selected"
)

// NormalizeRepositorySelection canonicalizes GitHub's repository_selection for
// storage. Empty means the caller didn't say, which is a supported state (the
// column is nullable and NULL reads back as ""). Anything outside GitHub's two
// values is refused rather than stored: the CHECK constraint would refuse it
// anyway, and failing here names the bad value instead of surfacing a driver
// error from three layers down.
func NormalizeRepositorySelection(selection string) (string, error) {
	s := strings.ToLower(strings.TrimSpace(selection))
	switch s {
	case "", RepositorySelectionAll, RepositorySelectionSelected:
		return s, nil
	default:
		return "", fmt.Errorf("unknown repository selection %q", selection)
	}
}

// Suspended reports whether the installation is currently suspended. The
// timestamp is the state (there is no separate flag to fall out of step with
// it), so every reader asks this rather than re-deriving the zero check.
func (i OrgGitHubAppInstallation) Suspended() bool { return !i.SuspendedAt.IsZero() }

// GrantsEveryRepository reports whether this installation's grant covers every
// repository on the account. Callers asking "can this installation drift?" ask
// this rather than comparing the string, and an unknown selection ("") answers
// false — the conservative side, since it renders as "we have not established
// that drift is impossible" rather than as an all-clear.
func (i OrgGitHubAppInstallation) GrantsEveryRepository() bool {
	return i.RepositorySelection == RepositorySelectionAll
}

// ReachableRepository is one entry in an org's reachable-repo cache — a
// repository the org's GitHub credentials can reach, whether or not any team
// tracks it.
//
// It is deliberately NOT a Repository: reachability is a cache of an external
// fact, rebuilt in full on every refresh, while a repository row is a registry
// of a TF entity that worktrees, entities and clone state hang off. A reachable
// entry with no registry row behind it is the normal case — under an App it is
// exactly what "reach without purpose" means, and under a PAT it is every
// repository the user can see but nobody tracks — so this type carries
// everything a reader needs without a join.
type ReachableRepository struct {
	OrgID string
	// CredentialClass is which credential system observed this reach. Every
	// read filters on the org's current class, so a tier switch cannot serve the
	// previous tier's answer alongside the new one.
	CredentialClass GitHubCredentialClass
	// InstallationID scopes a byo_app entry to the installation it was read
	// through; empty on a pat entry, which has no installation row to hang off.
	InstallationID string
	// Host scopes a pat entry to the GitHub deployment it was read against;
	// empty on a byo_app entry, which inherits the host from its installation.
	Host string
	// Source is the provider that issued the repository (RepoSourceGitHub
	// today); empty on a struct built without one, which the stores normalize on
	// write. It exists so the join to the registry is written against the same
	// key the registry is keyed under, not against a literal.
	Source string
	Owner  string
	Repo   string
	// ExternalID is GitHub's own repository id, which both enumerations return
	// for every entry. "" when the source sent none, which is a supported state:
	// the slug matches anyway.
	ExternalID string
	// The display fields the repository picker renders, mirrored from the same
	// repository objects the enumeration already returns.
	Description string
	Language    string
	HTMLURL     string
	// PushedAt is GitHub's own timestamp string, passed through verbatim. It is
	// never parsed or compared here.
	PushedAt string
	Private  bool
	// ObservedAt is when the refresh that wrote this row ran — display
	// provenance for "as of…" and the staleness gate the TTL and the write gate
	// both read.
	ObservedAt time.Time
}

// Slug renders the entry as the "owner/repo" string every slug-keyed column and
// store method uses.
func (r ReachableRepository) Slug() string { return r.Owner + "/" + r.Repo }

// ReachableCacheState is what a consumer needs to know about an org's cache
// without reading a single row of it: whether it has ever been refreshed, how
// much is in it, and how old the stalest part is.
//
// Refreshed and Count are separate on purpose, and conflating them is the bug
// this type exists to prevent. A scope that genuinely reaches NOTHING — an App
// installed on an account and then narrowed to no repositories — leaves zero
// rows behind, exactly like a refresh that never ran. Read off the row count,
// that org's picker would say "discovering repositories…" forever and kick a
// refresh on every open, which is the unbounded enumeration the TTL exists to
// prevent. Refreshed answers "have we looked"; Count answers "what did we find".
//
// ObservedAt is the OLDEST instant across the class's refreshed scopes, not the
// newest. A cache is only as fresh as its stalest scope: an org with two App
// installations, one refreshed a minute ago and one a week ago, is a week stale
// for the repositories on the second account, and taking the newest would report
// it fresh and never refresh the half that needs it.
type ReachableCacheState struct {
	Refreshed  bool
	Count      int
	ObservedAt time.Time
}

// Known reports whether any scope of this org and class has been refreshed. It
// is what the picker's readiness discriminator answers off — false is "we have
// not looked yet", never "you have no repositories".
func (s ReachableCacheState) Known() bool { return s.Refreshed }

// StaleAt reports whether the cache is older than ttl as of now. A cache nobody
// has refreshed is stale by definition — there is nothing to serve and nothing
// to date — so a caller that wants to distinguish the two checks Known first.
func (s ReachableCacheState) StaleAt(now time.Time, ttl time.Duration) bool {
	if !s.Refreshed || s.ObservedAt.IsZero() {
		return true
	}
	return now.Sub(s.ObservedAt) >= ttl
}

// GitHubPendingBind is one initiated bind ceremony: an admin clicked Connect,
// TF minted a nonce, and the browser left for GitHub carrying it in a cookie.
// The row is what the callback consumes to learn WHICH workspace the returning
// installation belongs to — the redirect from GitHub cannot say, because its
// installation_id is an unsigned, guessable query parameter GitHub's own
// documentation says not to trust.
//
// It is durable rather than in-process because the pod that serves the Connect
// click is not necessarily the pod that serves the callback. An in-memory map
// would work in development and fail intermittently under load, which is the
// worst failure shape a security control can have.
//
// NonceHash is the SHA-256 of the nonce in lowercase hex — never the nonce
// itself, so a database read yields nothing that can complete a bind. The
// browser's cookie holds the only copy of the plaintext.
type GitHubPendingBind struct {
	NonceHash string
	OrgID     string
	// UserID is the admin who started the ceremony. The callback requires the
	// returning session to be that same person, so a cookie replayed into
	// somebody else's browser consumes nothing.
	UserID    string
	CreatedAt time.Time
	ExpiresAt time.Time
	// ConsumedAt is when the callback spent this record, zero while it is
	// unspent. Consumption is a conditional update rather than a read then a
	// write, so two concurrent callbacks cannot both find it unspent.
	ConsumedAt time.Time
}

File: frontend/src/lib/githubApp.ts
// Helpers for the Workspace Settings "GitHub access" panel. The
// registration ceremony is a top-level browser navigation to a backend
// bounce page, which then POSTs the manifest to github.com (manifest
// flow) — see startGitHubAppRegistration. The endpoints are org-scoped
// under /api/orgs/{org_id}/github/app.

import { apiErrors, apiFetch, apiJSON, HttpError, httpErrorMessage } from './apiClient'
import { isHttpUrl } from './reachability'

// asError turns any failed call into an Error carrying the clean user-facing
// string — the server's `{ error }` verbatim (e.g. cutover's 409 "install the
// App before switching"), else the caller's fallback. Every function here is
// consumed by a panel that renders `err.message` straight into the UI, so a
// raw HttpError — whose message embeds the whole response body — must not
// escape this module.
function asError(e: unknown, fallback: string): Error {
  return new Error(httpErrorMessage(e, fallback))
}

// Local mode runs single-org against the synthetic sentinel org. Mirrors
// runmode.LocalDefaultOrgID on the backend; the org-scoped App endpoints
// take the id in the path even in local mode.
export const LOCAL_DEFAULT_ORG_ID = '00000000-0000-0000-0000-000000000001'

export interface GitHubAppInstallation {
  installation_id: string
  account_type: string
  account_login: string
  installed_at: string
  // RFC3339 when the account owner has suspended the installation, '' when it
  // is live: the grant survives a suspension, but GitHub refuses every token
  // minted from it. suspended_by is the login that suspended it ('' when
  // unsuspended, or when GitHub named no one). A suspended installation gets
  // its own visual state: one that merely looked connected would explain
  // nothing about the 403s every run under it earns.
  suspended_at: string
  suspended_by: string
  // Whether the grant is every repository on the account ('all') or an
  // enumerated set ('selected'), or null when the mirror has not learned it
  // yet. Three values, three sentences: an 'all' grant cannot be drifted out
  // of, a 'selected' one can, and null is "not known yet" — never either.
  repository_selection: 'all' | 'selected' | null
  // The installation's settings page on GitHub, where the grant is chosen. TF
  // links out and never edits the grant itself — GitHub enforces who may.
  settings_url: string
}

export interface GitHubAppInfo {
  app_id: string
  slug: string
  // 'user' (personal account) or 'org' (organization) — the account the App
  // was registered under, persisted at registration and used to seed the
  // "App account type" picker so it doesn't re-default to Personal on reload.
  owner_type: string
  registered_at: string
  registered_by_display_name: string
  // false while the registration is STAGED — registered during a PAT→App
  // switch but not yet cut over, so the PAT is still the live credential
  // (TFAC-328). The Setup/Settings UX reads this to resolve the live mode,
  // paint the "switch pending" mode-card state, and show the staged-switch
  // banner. true once a cutover activates it.
  active: boolean
}

// GitHubAppWebhookState is the backend's answer to "is GitHub actually
// delivering this App's webhooks here?", probed against the App's own webhook
// configuration and delivery history (never inferred from a stored secret,
// which is true in neither direction).
//
//   - not_configured: no webhook, or the inert placeholder a deployment GitHub
//     can't reach registers. The NORMAL state in local mode — not a fault.
//   - pointing_elsewhere: the App has a webhook, and it isn't this deployment.
//   - delivering_rejected: GitHub is delivering here and the receiver refuses.
//   - no_deliveries: configured for this deployment, nothing in GitHub's
//     30-day window. Ordinary for a new or quiet App.
//   - healthy: the most recent delivery was accepted.
//   - unavailable: this GitHub host doesn't expose the App-webhook endpoints
//     (GHES below 3.2), so the question can't be answered.
export type GitHubAppWebhookState =
  | 'not_configured'
  | 'pointing_elsewhere'
  | 'delivering_rejected'
  | 'no_deliveries'
  | 'healthy'
  | 'unavailable'

// GitHubAppWebhookHealth is the probe's answer as facts, not copy — the panel
// composes the message, because the likely cause of a rejection depends on the
// status code (a 401 reads differently from a failure to connect).
//
// secret_configured is GitHub's masked presence bit for the App's own webhook
// secret. It never carries the value, and TF never compares one: the delivery
// status codes are what prove a secret matches.
export interface GitHubAppWebhookHealth {
  state: GitHubAppWebhookState
  // Origin (scheme://host) the App delivers to, '' when nothing is configured.
  hook_host: string
  secret_configured: boolean
  // RFC3339 of the most recent delivery, '' when there is none.
  last_delivery_at: string
  // What the receiving endpoint answered on that delivery; 0 when GitHub could
  // not connect at all.
  last_delivery_status_code: number
  checked_at: string
}

export interface GitHubAppStatus {
  app: GitHubAppInfo | null
  installations: GitHubAppInstallation[]
  using_deployment_default: boolean
  // Whether this deployment offers a deployment App for a workspace to bind —
  // a fact about the deployment, not the org. It decides whether the
  // no-credential empty state offers Connect beside register, import and a
  // token; always false in local mode, where the managed class does not exist.
  deployment_app_available: boolean
  // null when there's no App, no deployment identity to compare a hook URL
  // against, or no probe answer yet. Absent means NOT KNOWN — never "fine".
  webhook_health: GitHubAppWebhookHealth | null
  // The absolute redirect_uri the App owner must register on the App for per-user
  // "Connect GitHub" OAuth to work. Same URL a manifest-created App already has
  // registered (harmless there); load-bearing for a bring-your-own App whose
  // owner registers it by hand. Empty when no deployment identity is configured.
  connect_callback_url: string
}

export async function getGitHubAppStatus(orgId: string): Promise<GitHubAppStatus> {
  try {
    return await apiJSON<GitHubAppStatus>(`/api/orgs/${encodeURIComponent(orgId)}/github/app`)
  } catch (e) {
    throw asError(e, 'Could not load GitHub App status.')
  }
}

// refreshGitHubAppInstallations reconciles the org's App installation mirror
// against GitHub on demand (TFAC-324's POST endpoint) and returns the refreshed
// status. This is the authoritative install-discovery path in local mode, where
// webhook deliveries never reach the NAT'd host — a plain getGitHubAppStatus
// reads a mirror a brand-new install hasn't touched yet, so the wizard's install
// step and Settings' "Check installation" action POST here to actually find it.
export async function refreshGitHubAppInstallations(orgId: string): Promise<GitHubAppStatus> {
  try {
    return await apiJSON<GitHubAppStatus>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/installations/refresh`,
      { method: 'POST' },
    )
  } catch (e) {
    throw asError(e, 'Could not refresh the GitHub App installations.')
  }
}

// GitHubWebhookReplayResult reports what the repair did. candidates is how many
// failed installation deliveries GitHub still holds in its 30-day window, which
// is what makes `replayed: 0` legible — nothing to replay reads very differently
// from nothing accepted.
export interface GitHubWebhookReplayResult {
  candidates: number
  replayed: number
  failed: number
}

// replayGitHubWebhookDeliveries asks GitHub to redeliver the App's failed
// installation deliveries, healing an installation mirror that went stale while
// the receiver was rejecting them (a missing or mismatched webhook secret). Only
// failed deliveries are replayed, and the receiver dedups on the delivery GUID a
// redelivery shares with its original, so this cannot double-apply anything.
//
// POST /api/orgs/{org_id}/github/app/webhook/replay
export async function replayGitHubWebhookDeliveries(
  orgId: string,
): Promise<GitHubWebhookReplayResult> {
  try {
    return await apiJSON<GitHubWebhookReplayResult>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/webhook/replay`,
      { method: 'POST' },
    )
  } catch (e) {
    throw asError(e, 'Could not replay the missed webhook deliveries.')
  }
}

export async function getGitHubAppInstallURL(orgId: string): Promise<string> {
  let body: { url: string }
  try {
    body = await apiJSON<{ url: string }>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/install-url`,
    )
  } catch (e) {
    throw asError(e, 'Could not load the install URL.')
  }
  // The URL is backend-derived from the org's GitHub base URL. Reject any
  // non-http(s) scheme here, at the single source, so neither an <a href>
  // (RepoPickerModal) nor window.open (Settings) can ever be handed a
  // javascript: link from a misconfigured base URL.
  if (!isHttpUrl(body.url)) {
    throw new Error('GitHub App install URL has an unsupported scheme')
  }
  return body.url
}

// ── Bring-your-own-App import ─────────────────────────────────────────────
// The second way into App mode: supply an App ID + private key for an App that
// already exists, validated server-side via an app-JWT GET /app round trip. The
// backend derives slug/owner/permissions from the App itself, permission-
// preflights, and persists through the same path the manifest callback uses.

// GitHubAppPermissionRow is one entry in the import preflight's granted-vs-
// required table. severity is 'required' (a hard gap blocks the import) or
// 'optional' (a soft gap blocks only until accept_partial acknowledges it);
// feature names the capability an unmet optional permission degrades.
export interface GitHubAppPermissionRow {
  permission: string
  required: string
  granted: string
  satisfied: boolean
  severity: string
  feature?: string
}

// GitHubAppImportInput is the import request body. client_secret (enables OAuth
// Connect), webhook_secret (multi mode only), and accept_partial (acknowledges
// soft permission gaps) are optional.
export interface GitHubAppImportInput {
  app_id: string
  pem: string
  client_secret?: string
  webhook_secret?: string
  accept_partial?: boolean
}

// GitHubAppImportResult is the success body: the status payload plus the
// preflight table and the client-secret validation outcome.
export interface GitHubAppImportResult extends GitHubAppStatus {
  permissions: GitHubAppPermissionRow[]
  client_secret_stored: boolean
  client_secret_validated: boolean
}

// GitHubAppImportOutcome discriminates a successful import from a rejection. On
// failure, `permissions` + `blocking` are present for a permission gap (blocking
// = a hard gap accept_partial can't override; !blocking = a soft gap the caller
// resubmits past with accept_partial); `field` is present for a field-level
// rejection (bad PEM, app-id mismatch, bad client secret).
export type GitHubAppImportOutcome =
  | { ok: true; result: GitHubAppImportResult }
  | {
      ok: false
      error: string
      permissions?: GitHubAppPermissionRow[]
      blocking?: boolean
      field?: string
    }

// importGitHubApp validates and imports an existing GitHub App. It returns a
// structured outcome rather than throwing, so the form can render the permission
// table on a gap (instead of a bare error string).
//
// POST /api/orgs/{org_id}/github/app/import
export async function importGitHubApp(
  orgId: string,
  input: GitHubAppImportInput,
): Promise<GitHubAppImportOutcome> {
  const fallback = 'Could not import the GitHub App.'
  try {
    const result = await apiJSON<GitHubAppImportResult>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/import`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(input),
      },
    )
    return { ok: true, result }
  } catch (e) {
    // Every rejection carries its message in the standard envelope, read
    // through the shared parser so the field attribution comes with it. A
    // permission gap adds to that rather than replacing it: alongside `errors`
    // its body carries the granted-vs-required table the form renders, which no
    // envelope field can hold (see githubAppImportErrorResponse). So check for
    // the table first — the message is the same one either way.
    const item = apiErrors(e)[0]
    if (e instanceof HttpError) {
      try {
        const body = JSON.parse(e.body) as {
          permissions?: GitHubAppPermissionRow[]
          blocking?: boolean
        }
        if (body.permissions) {
          return {
            ok: false,
            error: item?.message || fallback,
            permissions: body.permissions,
            blocking: body.blocking,
          }
        }
      } catch {
        // Body wasn't JSON — fall through.
      }
    }
    if (item) return { ok: false, error: item.message || fallback, field: item.field }
    return { ok: false, error: httpErrorMessage(e, fallback) }
  }
}

// startGitHubAppRegistration kicks off the manifest flow with a top-level
// navigation to the backend launch endpoint. The SPA's global CSP sets
// `form-action 'self'`, which would block an in-page form POST to
// github.com — but it does NOT govern top-level navigation. The launch
// page then renders the manifest form under its own per-response CSP
// (scoped to the org's GitHub host) and does the cross-origin POST when
// the user clicks "Continue to GitHub". GitHub redirects back to the
// callback, which lands the browser back where registration was launched
// from: `return_to` ('setup' from the wizard, 'settings' from Settings) is
// threaded through the signed state so the user resumes where they left off
// rather than always being dropped on Settings.
export function startGitHubAppRegistration(
  orgId: string,
  payload: { owner_type: 'user' | 'org'; owner_login: string; return_to: 'setup' | 'settings' },
): void {
  const params = new URLSearchParams({
    owner_type: payload.owner_type,
    owner_login: payload.owner_login,
    return_to: payload.return_to,
  })
  window.location.assign(
    `/api/orgs/${encodeURIComponent(orgId)}/github/app/register/launch?${params.toString()}`,
  )
}

// ── PAT ↔ App switching (TFAC-328 endpoints) ──────────────────────────────
// These drive the guided switch flows in the wizard and Settings. The backend
// enforces the either/or invariant; the UI's job is to preview the consequence
// (the reachability diff), confirm, and commit.

// A repo the target credential can't reach after a switch, with the teams that
// track it — what the inform-only diff lists so an admin knows who's affected.
export interface DarkRepo {
  repo: string
  teams: string[]
}

// The inform-only reachability diff both switch preflights return: how many of
// the org's tracked repos the target credential reaches, and which go dark.
export interface AccessDiff {
  tracked: number
  reachable: number
  dark_repos: DarkRepo[]
}

// pat-preflight's body — the diff plus the login the PAT authenticates as.
export interface PatPreflight extends AccessDiff {
  login: string
}

// switch-to-pat's success body. github_app_settings_url points at the org
// host's apps page so the UI can guide deleting the (still-existing) App on
// GitHub; github_app_deleted_locally is always true on success (the local
// registration + secrets are gone).
export interface SwitchToPatResult {
  status: string
  github_app_deleted_locally: boolean
  github_app_settings_url: string
}

// cutoverToApp commits a staged PAT→App switch: the backend reconciles
// installations, verifies ≥1, then activates the App and deletes the PAT
// atomically. Rejects with the backend's message — notably the 409 "install
// the App before switching" when no installation is found yet.
//
// POST /api/orgs/{org_id}/github/app/cutover
export async function cutoverToApp(orgId: string): Promise<void> {
  try {
    await apiFetch(`/api/orgs/${encodeURIComponent(orgId)}/github/app/cutover`, {
      method: 'POST',
    })
  } catch (e) {
    throw asError(e, 'Could not switch to the GitHub App.')
  }
}

// cutoverPreflight returns the inform-only reachability diff for a PAT→App
// cutover (which tracked repos the App's installations would leave dark). Has a
// server-side write side-effect (it reconciles the installation mirror), so the
// endpoint is uncacheable — call it right before showing the diff screen.
//
// GET /api/orgs/{org_id}/github/app/cutover-preflight
export async function cutoverPreflight(orgId: string): Promise<AccessDiff> {
  try {
    return await apiJSON<AccessDiff>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/cutover-preflight`,
    )
  } catch (e) {
    throw asError(e, 'Could not check repository reachability.')
  }
}

// switchToPat performs the App→PAT switch — a full App teardown. The backend
// validates the PAT, stores it, and deletes the App registration + secrets in
// one transaction; the App still exists on GitHub (the result flags that).
//
// POST /api/orgs/{org_id}/github/pat/switch-to
export async function switchToPat(orgId: string, pat: string): Promise<SwitchToPatResult> {
  try {
    return await apiJSON<SwitchToPatResult>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/pat/switch-to`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ pat }),
      },
    )
  } catch (e) {
    throw asError(e, 'Could not switch to a personal access token.')
  }
}

// patPreflight validates a candidate PAT and returns the inform-only
// reachability diff for an App→PAT switch plus the login it authenticates as.
// Stores nothing — switchToPat re-validates on commit.
//
// POST /api/orgs/{org_id}/github/pat/preflight
export async function patPreflight(orgId: string, pat: string): Promise<PatPreflight> {
  try {
    return await apiJSON<PatPreflight>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/pat/preflight`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ pat }),
      },
    )
  } catch (e) {
    throw asError(e, 'That token could not be validated.')
  }
}

// discardStagedApp tears down a STAGED App registration — the exit for an
// abandoned PAT→App switch. The live PAT is untouched. 409s for an active App
// (disconnectOwnApp removes a live one).
//
// DELETE /api/orgs/{org_id}/github/app
export async function discardStagedApp(orgId: string): Promise<void> {
  try {
    await apiFetch(`/api/orgs/${encodeURIComponent(orgId)}/github/app`, { method: 'DELETE' })
  } catch (e) {
    throw asError(e, 'Could not discard the staged GitHub App.')
  }
}

// disconnectOwnApp tears down the workspace's LIVE App with nothing bound in
// its place: registration, installations, secrets and host go in one
// transaction and the workspace is left as a fresh one, with every way in
// open — the deployment's App included, which refuses a workspace that still
// holds a credential. The App still exists on GitHub (the result carries the
// link out, as switch-to-pat's does). 409s for a staged App (discardStagedApp
// is that door).
//
// POST /api/orgs/{org_id}/github/app/disconnect
export async function disconnectOwnApp(orgId: string): Promise<SwitchToPatResult> {
  try {
    return await apiJSON<SwitchToPatResult>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/app/disconnect`,
      { method: 'POST' },
    )
  } catch (e) {
    throw asError(e, 'Could not disconnect the GitHub App.')
  }
}

// ── The deployment App (managed class, multi mode only) ───────────────────

// startManagedGitHubConnectAccount starts the bind ceremony — the one start
// there is. The admin names the account; the backend mints the pending-bind
// record and cookie and answers with GitHub's OAuth authorize URL, which this
// navigates to. GitHub proves who the admin is and returns them to the
// callback, where the installation is found among the ones that person can
// see and the bind completes — or, for an account that has no installation
// they can see, the callback sends them on to GitHub's install page with the
// account preselected and completes when GitHub returns. Nothing is listed
// and nothing is picked from: the name is the admin's, the proof is GitHub's.
// Control never comes back here on success; a refusal (the deployment has no
// App, the workspace already holds a credential) rejects with the backend's
// sentence.
//
// POST /api/orgs/{org_id}/github/managed/connect-account
export async function startManagedGitHubConnectAccount(
  orgId: string,
  account: string,
): Promise<void> {
  let authorizeUrl: string
  try {
    const out = await apiJSON<{ authorize_url: string }>(
      `/api/orgs/${encodeURIComponent(orgId)}/github/managed/connect-account`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ account: account.trim() }),
      },
    )
    authorizeUrl = out.authorize_url
  } catch (e) {
    throw asError(e, 'Could not start connecting that account.')
  }
  window.location.assign(authorizeUrl)
}

// disconnectManagedGitHub moves the workspace off the deployment App: every
// bound installation is released and the credential class resets, so the
// workspace is back in its no-credential state and free to register its own
// App or bind a token. Nothing is uninstalled on GitHub — the installation
// persists there unbound, and startManagedGitHubConnectAccount re-binds it.
//
// POST /api/orgs/{org_id}/github/managed/disconnect
export async function disconnectManagedGitHub(orgId: string): Promise<void> {
  try {
    await apiFetch(`/api/orgs/${encodeURIComponent(orgId)}/github/managed/disconnect`, {
      method: 'POST',
    })
  } catch (e) {
    throw asError(e, 'Could not disconnect the deployment GitHub App.')
  }
}

// disconnectManagedInstallation releases one bound account and keeps the
// class — unless it was the last one, in which case it is the full disconnect.
// The same verb narrowed, so the panel's per-account button and its Disconnect
// button can never leave the bindings and the class disagreeing.
//
// POST /api/orgs/{org_id}/github/managed/installations/{installation_id}/disconnect
export async function disconnectManagedInstallation(
  orgId: string,
  installationId: string,
): Promise<void> {
  try {
    await apiFetch(
      `/api/orgs/${encodeURIComponent(orgId)}/github/managed/installations/${encodeURIComponent(installationId)}/disconnect`,
      { method: 'POST' },
    )
  } catch (e) {
    throw asError(e, 'Could not disconnect that GitHub account.')
  }
}

// ── The two grant findings ────────────────────────────────────────────────
// Computed server-side from the reachable-repo mirror, never by asking GitHub
// on page load, and served as ordinary paginated lists (usePagedList over the
// paths below). Both exist only for an App-class workspace — a PAT's reach is
// not a grant TF holds, and the routes 404 for one.

// A repository the App can reach that no team tracks: TF holds write access to
// code nobody asked it to touch. The verb is on GitHub — narrow the grant on
// the installation's settings page (settings_url), or uninstall.
export interface ReachWithoutPurposeItem {
  installation_id: string
  account_login: string
  settings_url: string
  owner: string
  repo: string
  slug: string
  private: boolean
  html_url: string
  observed_at: string
}

// A repository some team tracks that the grant does not contain, so it is
// silently unpolled. installation_id names the installation on the owner
// account when there is one (by construction a 'selected' grant, since an
// 'all' grant cannot drift and an unknown one is not reported), '' when no
// bound installation covers that account at all.
export interface ScopeDriftItem {
  owner: string
  repo: string
  slug: string
  installation_id: string
  account_login: string
  settings_url: string
}

export function reachWithoutPurposeListPath(orgId: string): string {
  return `/api/orgs/${encodeURIComponent(orgId)}/github/grant/reach-without-purpose/list`
}

export function scopeDriftListPath(orgId: string): string {
  return `/api/orgs/${encodeURIComponent(orgId)}/github/grant/scope-drift/list`
}
